Source/WebCore/platform/text/TextSpacing.cpp

```diff
@@ -0,0 +1,111 @@
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "TextSpacing.h"
+
+namespace WebCore {
+
+using namespace TextSpacing;
+
+bool TextAutospace::shouldApplySpacing(char32_t firstCharacter, char32_t secondCharacter) const
+{
+    auto firstCharacterClass = characterClass(firstCharacter);
+    auto secondCharacterClass = characterClass(secondCharacter);
+
+    if (hasIdeographAlpha()) {
+        if (firstCharacterClass == CharacterClass::Ideograph)
+            return secondCharacterClass == CharacterClass::NonIdeographLetter;
+        if (firstCharacterClass == CharacterClass::NonIdeographLetter && secondCharacterClass == CharacterClass::Ideograph)
+            return secondCharacterClass == CharacterClass::Ideograph;
+    }
+    return false;
+}
+namespace TextSpacing {
+
+bool isIdeograph(char32_t character)
+{
+    // Lowest possible ideographic codepoint
+    if (character < 0x2E80)
+        return false;
+
+    // All characters in the range of U+3041 to U+30FF, except those that belong to Unicode Punctuation [P*]
+    if ((character >= 0x3041 && character <= 0x30FF) && !isPunctuation(character))
+        return true;
+    // CJK Strokes (U+31C0 to U+31EF).
+    if (character >= 0x31C0 && character <= 0x31EF)
+        return true;
+    // Katakana Phonetic Extensions (U+31F0 to U+31FF)
+    if (character >= 0x31F0 && character <= 0x31FF)
+        return true;
+    if (isOfScriptType(character, UScriptCode::USCRIPT_HAN))
+        return true;
+
+    return false;
+}
+
+static bool isNonIdeographicNumeral(char32_t character, uint32_t generalCategoryMask)
+{
+    // FIXME: Should also check that it is not: upright in vertical text flow using the text-orientation property or the text-combine-upright property.
+    return ((generalCategoryMask & U_GC_ND_MASK) && !isEastAsianFullWidth(character));
+}
+
+// Classes are defined at https://www.w3.org/TR/css-text-4/#text-spacing-classes
+CharacterClass characterClass(char32_t character)
+{
+    auto generalCategoryMask = U_GET_GC_MASK(character);
+    if (isIdeograph(character))
+        return CharacterClass::Ideograph;
+
+    // We already know it is not an Ideograph from here
+    // FIXME: Should also check that it is not: upright in vertical text flow using the text-orientation property or the text-combine-upright property.
+    if (generalCategoryMask & (U_GC_M_MASK | U_GC_L_MASK)) {
+        if (!isEastAsianFullWidth(character))
+            return CharacterClass::NonIdeographLetter;
+        // General Category M/L won't apply to anything else
+        return CharacterClass::Other;
+    }
+
+    if (isNonIdeographicNumeral(character, generalCategoryMask))
+        return CharacterClass::NonIdeographNumeral;
```

Comment:
We already know it's not ideographic, right? So shouldn't this just check if it's a numeral?

Reply:
I'm not sure...Spec says:

> non-ideographic numerals
> Includes all [typographic character](https://www.w3.org/TR/css-text-4/#typographic-character-unit) units that belong to the Unicode Decimal Digit Number [Nd] [general category](https://www.w3.org/TR/css-text-4/#unicode-general-category), except when any of the following conditions are met:
> is categorized as East Asian Fullwidth (F) by [[UAX11]](https://www.w3.org/TR/css-text-4/#biblio-uax11).
> is upright in vertical text flow using the [text-orientation](https://www.w3.org/TR/css-writing-modes-4/#propdef-text-orientation) property or the [text-combine-upright](https://www.w3.org/TR/css-writing-modes-4/#propdef-text-combine-upright) property.

So I guess we can't do that unless we can guarantee that a character that is not a ideograph can't be classified as East Asian fullwidth.

Reply:
Ah, right.